odoo/orm/domains.py

```diff
@@ -631,6 +655,21 @@ def __and__(self, other):
             return DomainAnd(self.children + other.children)
         return super().__and__(other)
 
+    def _filter_records(self, records):
+        predicates = []
+
+        def filter_and(record):
+            if not all(pred(record) for pred in predicates):
+                return False
+            for index in range(len(predicates), len(self.children)):
+                child = self.children[index]
+                predicates.append(pred := child._filter_records(records))
+                if not pred(record):
+                    return False
+            return True
```

Comment:
Wow, the function name is confusing and this code is way too subtle.  I really am struggling to understand how it works.  Please provide a formal proof of its correctness, or rewrite it with plain stupid code, like
```py
def predicate_and(record):
    if not predicates:  # lazy computation of predicates
        predicates.extend(child._filter_records(records) for child in self.children)
    return all(predicate(record) for predicate in predicates)
```
Same remark for `filter_or`.

Reply:
Comments added. In the current version, predicates are lazily computed one by one (not all at the same time).